test/UpdaterAndFilterEngineCreation.cpp

```diff
@@ -0,0 +1,172 @@
+      threadsVector.reserve(COUNT);
+      std::uninitialized_fill(updaterAddrArray, updaterAddrArray + COUNT, nullptr);
+      std::uninitialized_fill(filterAddrArray, filterAddrArray + COUNT, nullptr);
+    }
+
+    void CallGetUpdaterSimultaneously()
+    {
+      CallGetSimultaneously(true, false);
+    }
+
+    void CallGetFilterEngineSimultaneously()
+    {
+      CallGetSimultaneously(false, true);
+    }
+
+    void CallGetUpdaterAndGetFilterEngineSimultaneously()
+    {
+      CallGetSimultaneously(true, true);
+    }
+
+  private:
+
+    void CallGetSimultaneously(bool isUpdater, bool isFilterEngine)
+    {
+      for (size_t idx = 0; idx < COUNT; ++idx)
+        threadsVector.push_back(
+          std::thread([this, idx, isUpdater, isFilterEngine]() -> void {
+            Updater* updaterAddr = nullptr;
+            FilterEngine* filterEngineAddr = nullptr;
+            if (isUpdater && isFilterEngine)
+            {
+              // some randomization in order of calling gets
+              if (idx % 2)
+              {
+                updaterAddr = &(platform->GetUpdater());
+                filterEngineAddr = &(platform->GetFilterEngine());
+              }
+              else
+              {
+                filterEngineAddr = &(platform->GetFilterEngine());
+                updaterAddr = &(platform->GetUpdater());
+              }
+            }
+            else if (isUpdater)
+              updaterAddr = &(platform->GetUpdater());
+            else if (isFilterEngine)
+              filterEngineAddr = &(platform->GetFilterEngine());
+
+            if (updaterAddr != nullptr)
+              updaterAddrArray[idx] = updaterAddr;
+
+            if (filterEngineAddr != nullptr)
+              filterAddrArray[idx] = filterEngineAddr;
+          }));
+
+      // join the threads
+      for (auto& th: threadsVector)
+        if (th.joinable())
+          th.join();
+    }
+  };
+}
+
+TEST_F(UpdaterAndFilterEngineCreationTest, TestFilterEngineSingleInstance)
+{
+    CallGetFilterEngineSimultaneously();
+    FilterEngine* filterEngineAddr = filterAddrArray[0];
+    for (size_t i = 1; i < COUNT; ++i)
+      ASSERT_EQ(filterEngineAddr, filterAddrArray[i]);
+}
+
+TEST_F(UpdaterAndFilterEngineCreationTest, TestUpdaterSingleInstance)
+{
+    CallGetUpdaterSimultaneously();
+    Updater* updaterAddr = updaterAddrArray[0];
+    for (size_t i = 1; i < COUNT; ++i)
+      ASSERT_EQ(updaterAddr, updaterAddrArray[i]);
+}
+
+TEST_F(UpdaterAndFilterEngineCreationTest, TestUpdaterAndFilterEngineCreationsDontCollide)
+{
+    CallGetUpdaterAndGetFilterEngineSimultaneously();
+    ASSERT_EQ(filterAddrArray[0], filterAddrArray[COUNT - 1]);
+    ASSERT_EQ(updaterAddrArray[0], updaterAddrArray[COUNT - 1]);
+}
+
+TEST_F(UpdaterAndFilterEngineCreationTest, TestUpdaterAndFilterEngineCreationOrder1)
+{
+    Updater& updater = platform->GetUpdater();
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Just wonder, does it not work without that sleep?

Reply:
No, sleep is not needed.